### PR TITLE
Add validator for repo id + decorator to validate arguments in `huggingface_hub`

### PR DESCRIPTION
**Pull request by Wauplin**
Discussed in https://github.com/huggingface/huggingface_hub/issues/1008.

The goal is to avoid receiving repo ids in the form `datasets/user/repo_name` or as a Path and not noticing it (was the root cause of https://github.com/huggingface/transformers/issues/18691 but found out about it only after digging).

In addition, I tried to set up a small structure to harmonize the validation of other arguments that are used everywhere without duplicating the validation (`repo_type` for example). But for now, let's stick to `repo_id` only.

---
**EDIT**: just when opening this PR, I realized the `"datasets/huggingface/label-files"` form for `repo_id` is used ~25 times in the `transformers` library ([here in the code](https://github.com/huggingface/transformers/blob/ab663b227490ee82fe944bbf3680d26fd53b337c/src/transformers/models/beit/convert_beit_unilm_to_pytorch.py#L179) or [here in the examples](https://github.com/huggingface/transformers/blob/ab663b227490ee82fe944bbf3680d26fd53b337c/examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py#L390)). It's a bit annoying so I guess we will have to go through a deprecation cycle for that as well. Luckily, the form is used only for calls to `hf_hub_url` and `hf_hub_download`.
My suggestion would be to:
1. remove the `repo_id` validation for `hf_hub_url` and `hf_hub_download` for now
2. add a deprecation warning
3. fix `transformers` in the meantime to avoid "datasets/***" patterns
4. add `repo_id` validation in ~v0.12 to `hf_hub_url` and `hf_hub_download`

### CODE REVIEW

File: docs/source/package_reference/utilities.mdx [hunk 1]
```diff
@@ -60,3 +60,54 @@ See below for all custom errors thrown by different methods across the package.
...
+
+## Validators
+
+`huggingface_hub` include custom validators to validate method arguments automatically.
+Validation is inspired by the work done in [Pydantic](https://pydantic-docs.helpmanual.io/)
+to validate type hints but with more limited features.
+
+### Generic decorator
+
+[`~huggingface_hub.utils.validate_hf_hub_args`] is a generic decorator to encapsulate
```
Comment (sgugger) [inline docs/source/package_reference/utilities.mdx:72]:
```suggestion
[`~utils.validate_hf_hub_args`] is a generic decorator to encapsulate
```

File: docs/source/package_reference/utilities.mdx [hunk 2]
```diff
@@ -60,3 +60,54 @@ See below for all custom errors thrown by different methods across the package.
...
+methods that have arguments following `huggingface_hub`'s naming. By default, all
+arguments that has a validator implemented will be validated.
+
+If an input is not valid, a [`~huggingface_hub.utils.HFValidationError`] is thrown. Only
```
Comment (sgugger) [inline docs/source/package_reference/utilities.mdx:76]:
```suggestion
If an input is not valid, a [`~utils.HFValidationError`] is thrown. Only
```

**Review by sgugger — APPROVED** [review]
Left some nits on the docstrings, but otherwise looks good to me! Makes a lot of sense and we should update the uses in Transformers.

File: src/huggingface_hub/utils/_validators.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    avoid local inconsistencies whenever possible (example: passing `repo_type` in the
+    `repo_id` is forbidden).
+
+    Rules:
+    - Between 1 and 96 characters.
+    - Either "repo_name" or "namespace/repo_name"
+    - [a-zA-Z0-9] or "-", "_", "."
+    - "--" and ".." are forbidden
+
+    Examples:
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply src/huggingface_hub/utils/_validators.py]: Updated it to address that :)

File: src/huggingface_hub/utils/_validators.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    >>> my_cool_method("other..repo..id")
+    huggingface_hub.utils._validators.HFValidationError: Cannot have -- or .. in repo_id: 'other..repo..id'.
+
+    >>> my_cool_method(repo_id="other..repo..id")
+    huggingface_hub.utils._validators.HFValidationError: Cannot have -- or .. in repo_id: 'other..repo..id'.
+    ```
+
+    <Tip warning={true}>
+        Raises:
+            - [`~utils.HFValidationError`]: If an input is not valid.
```
Comment (Wauplin) [inline src/huggingface_hub/utils/_validators.py:73]:
```suggestion
    Raises:
        - [`~utils.HFValidationError`]: If an input is not valid.
```

File: docs/source/package_reference/utilities.mdx [hunk 5]
```diff
@@ -60,3 +60,54 @@ See below for all custom errors thrown by different methods across the package.
...
+
+If an input is not valid, a [`~utils.HFValidationError`] is thrown. Only
+the first non-valid value throws an error and stops the validation process.
+
+Usage:
+
+```py
+>>> from huggingface_hub.utils import validate_hf_hub_args
+
+>>> @validate_hf_hub_args
```
Comment (LysandreJik) [inline docs/source/package_reference/utilities.mdx:84]: Very very nice decorator, this should be super useful around the codebase and in third-party libraries.

File: src/huggingface_hub/utils/_validators.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    Example:
+    ```py
+    >>> from huggingface_hub.utils import validate_repo_id
+    >>> validate_repo_id(repo_id="valid_repo_id")
+    >>> validate_repo_id(repo_id="other..repo..id")
+    huggingface_hub.utils._validators.HFValidationError: Cannot have -- or .. in repo_id: 'other..repo..id'.
+    ```
+
+    Discussed in https://github.com/huggingface/huggingface_hub/issues/1008.
+    In moon-landing:
```
Comment (LysandreJik) [inline src/huggingface_hub/utils/_validators.py:121]:
```suggestion
    In moon-landing (internal repository):
```

**Review by LysandreJik — APPROVED** [review]
Looks good to me, thanks for working on it @Wauplin

**Comment by Wauplin** [comment]
Thanks both for the reviews and approval. It's merged :D

@sgugger how would you like to proceed to adapt the use cases in transformers (as described in https://github.com/huggingface/huggingface_hub/pull/1029#issue-1358903151). Should we create a separate issue to track that work ?

**Comment by sgugger** [comment]
Can you open a PR switching them all? That would be the easiest.

**Pull request merged**